### PR TITLE
Hyperswarm support

### PR DESCRIPTION
**Pull request by jeffpeterson**
A refactoring of hypermerge's networking to support hyperswarm.

### CODE REVIEW

File: src/MessageChannel.ts [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+export default class MessageChannel<Msg> {
+  stream: Duplex
+  sendQ: Queue<Msg>
+  receiveQ: Queue<Msg>
+
+  constructor(stream: Duplex) {
+    this.stream = stream
+
+    this.sendQ = new Queue('MessageBus:sendQ')
```
Comment (mjtognetti) [inline src/MessageChannel.ts]: `MessageChannel:*`

File: src/Metadata.ts [hunk 2]
```diff
@@ -271,10 +268,6 @@ export class Metadata {
     this.readyQ.subscribe((f) => f())
   }
 
-  private hasBlock(block: MetadataBlock): boolean {
```
Comment (mjtognetti) [inline src/Metadata.ts]: sad to see this go :(

File: src/Network.ts [hunk 3]
```diff
@@ -1,104 +1,109 @@
...
-      live: true,
-      id: decodePeerId(this.selfId),
-      encrypt: false,
-      timeout: 10000,
-      extensions: DocumentBroadcast.SUPPORTED_EXTENSIONS,
+  getOrCreatePeer(peerId: PeerId) {
+    return getOrCreate(this.peers, peerId, () => {
+      const peer = new NetworkPeer(this.selfId, peerId)
+
+      peer.connectionQ.subscribe((_conn) => {
```
Comment (mjtognetti) [inline src/Network.ts]: Trying to think of where the right place to document this behavior is, "We only present a peer after an initial connection has been established." Inline comments here don't really meet the need, which is "what are the characteristics of this thing?".

File: src/Network.ts [hunk 4]
```diff
@@ -1,104 +1,109 @@
...
+
+      return peer
     })
+  }
 
-    const onFeedRequested = (discoveryKey: Buffer) => {
-      const discoveryId = encodeDiscoveryId(discoveryKey)
-      const feedId = this.joined.get(discoveryId)
+  private onConnection = async (socket: Socket, details: ConnectionDetails) => {
+    details.reconnect(false)
```
Comment (mjtognetti) [inline src/Network.ts]: Are we still seeing hyperswarm attempting to reconnect over and over?

File: src/NetworkPeer.ts [hunk 5]
```diff
@@ -1,19 +1,104 @@
...
+  }
+
+  closeConnection(conn: PeerConnection): void {
+    conn.close()
+    this.closedConnectionCount += 1
+  }
+
+  close(): void {
+    this.connection && this.closeConnection(this.connection)
+  }
```
Comment (mjtognetti) [inline src/NetworkPeer.ts]: Is it possible to get into a state where we call `close` before `confirmConnection` and we have some pending connections in `pendingConnections` which aren't closed?

File: src/PeerConnection.ts [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+import { NetworkMsg } from './NetworkMsg'
+import pump from 'pump'
+
+export interface SocketInfo {
+  type: 'tcp' | 'utp'
+  isClient: boolean
+}
+
+export default class PeerConnection {
+  networkChannel: MessageChannel<NetworkMsg>
```
Comment (mjtognetti) [inline src/PeerConnection.ts]: If the multiplex streams are also called channels, I see your affinity for `MessageBus` here: `networkBus`.

Comment (mjtognetti) [reply src/PeerConnection.ts]: or even `networkMessageBus`

File: src/Network.ts (hunk 4 above)
Comment (jeffpeterson) [reply src/Network.ts]: Yeah :/

File: src/ReplicationManager.ts [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+  getPeersWith(discoveryIds: DiscoveryId[]): Set<NetworkPeer> {
+    return joinSets(discoveryIds.map((id) => this.peersByDiscoveryId.get(id)))
+  }
+
+  close(): void {
+    this.messages.inboxQ.unsubscribe()
+  }
+
+  /**
+   * Should be called when a peer connects.
```
Comment (mjtognetti) [inline src/ReplicationManager.ts]: Is there a possibility of this not being called?

File: src/types/multiplex.d.ts [hunk 8]
```diff
@@ -0,0 +1,54 @@
+declare module 'multiplex' {
```
Comment (mjtognetti) [inline src/types/multiplex.d.ts]: These type declaration files are bombdotcom

File: src/ReplicationManager.ts (hunk 7 above)
Comment (jeffpeterson) [reply src/ReplicationManager.ts]: Nah, just poor phrasing

**Pull request merged**